src/test/java/se/liu/ida/hefquin/engine/data/impl/VocabularyMappingTest.java

```diff
@@ -74,6 +74,226 @@ public void TranslateTriplePatternTest() {
+				
+		//Multiple mappings for same subject
+		final Node o2 = NodeFactory.createLiteral("s3");
+		testTriples.add(new Triple(s, p, o2));
+		
+		final VocabularyMapping vm = new VocabularyMappingImpl(testTriples);
+		
+		p = NodeFactory.createLiteral("p");
+		final Node o = NodeFactory.createLiteral("o");
+		final TriplePattern testTp = new TriplePatternImpl(s, p, o);
+		final SPARQLGraphPattern translation = vm.translateTriplePattern(testTp); 
+		final Set<Triple> translationTriples = new HashSet<>();
+		assertTrue(translation instanceof SPARQLUnionPatternImpl);
+		for (final SPARQLGraphPattern i : ((SPARQLUnionPatternImpl) translation).getSubPatterns()) {
+			assertTrue(i instanceof TriplePattern);
+			translationTriples.add(((TriplePattern) i).asJenaTriple());
+		}
+			
+		final Set<Triple> expectedResults = new HashSet<>();
+		expectedResults.add(new Triple(o1, p, o));
+		expectedResults.add(new Triple(o2, p, o));
+		
+		assertEquals(translationTriples, expectedResults);
+	}
+	
+	@Test
+	public void TranslatePredicateTest() {
+		final Set<Triple> testTriples = new HashSet<>();
+		//Equality
+		Node s = NodeFactory.createLiteral("p1");
+		Node p = OWL.equivalentProperty.asNode();
+		final Node t1 = NodeFactory.createLiteral("p2");
+		testTriples.add(new Triple(s, p, t1));
+				
+		//Predicate inverse
+		p = OWL.inverseOf.asNode();
+		final Node t2 = NodeFactory.createLiteral("Not p1");
+		testTriples.add(new Triple(s, p, t2));
+			
+		//Predicate subProperty
+		final Node t3 = NodeFactory.createLiteral("Subtype");
+		p = RDFS.subPropertyOf.asNode();
+		testTriples.add(new Triple(t3, p, s));
+		
+		//Predicate Intersection
+		p = OWL.equivalentProperty.asNode();
+		final Node t4 = NodeFactory.createBlankNode();
+		testTriples.add(new Triple(s, p, t4));	
+		p = OWL.intersectionOf.asNode();
+		final Node t5 = NodeFactory.createLiteral("p3");
+		testTriples.add(new Triple(t4, p, t5));		
+		final Node t6 = NodeFactory.createLiteral("p4");
+		testTriples.add(new Triple(t4, p, t6));
+		
+		//Predicate Union
+		p = OWL.equivalentProperty.asNode();
+		final Node t7 = NodeFactory.createBlankNode();
+		testTriples.add(new Triple(s, p, t7));	
+		p = OWL.unionOf.asNode();
+		final Node t8 = NodeFactory.createLiteral("p5");
+		testTriples.add(new Triple(t7, p, t8));		
+		final Node t9 = NodeFactory.createLiteral("p6");
+		testTriples.add(new Triple(t7, p, t9));
+
+		final VocabularyMapping vm = new VocabularyMappingImpl(testTriples);
+		
+		p = s;
+		s = NodeFactory.createLiteral("s");
+		final Node o = NodeFactory.createLiteral("o");
+		final TriplePattern testTp = new TriplePatternImpl(s, p, o);
+		final SPARQLGraphPattern translation = vm.translateTriplePattern(testTp); 
+		final Set<Triple> translationTriples = new HashSet<>();
+		assertTrue(translation instanceof SPARQLUnionPatternImpl);
+		boolean oneUnion = false;
+		boolean oneIntersection = false;
+		for (final SPARQLGraphPattern i : ((SPARQLUnionPatternImpl) translation).getSubPatterns()) {
```

Comment:
I'm a bit skeptic with this way of testing. Firstly, it seems to me that it checks for "at least one" union and intersection, not exactly one. On the other hand, I think it should be enough to test the direct object equality between the result of the translation and the expected result: `assertEquals(expectedPattern, translation)`, for which you need to instanciate `expectedPattern` in such a way that it contains the correct structure. For instance, like this:

```
final SPARQLGraphPattern expectedPattern = new SPARQLUnionPatternImpl(
   new TriplePattern(...), new TriplePattern(...), ...
);
```

Reply:
I have tried comparing SPARQLUnionPatterns, my issue is that they do not count as equal even if the members are equal. I have tried fixing this in the equals functions but I do not get it to work. The only comparisons that I've managed to make work are the ones using Jena Triples.

Reply:
I did test so that more than one union/intersection will not be accepted.

Reply:
I think that it would be best to investigate why the equals between SPARQLUnionPatternImpl doesn't work correctly. A first step is to implement the `toString` method, so that jUnit shows where the differences between the object are.

Reply:
I have pushed a version of the SPARQLUnionPattern equals function, where if the prints are used you can see that the Set contain the same memebers. I made them to sets, as you said earlier that lists take order into consideration. But even then they are not considered equal. I will keep looking at the equals functions of other classes to see where it goes wrong but I've been a bit stuck on this

Reply:
Give me a sec and I'll add some code to help, please don't push anything more yet.

Reply:
It seems to me that the order changes, even after changing the order of the expected elements they are still wrong. 

When it comes to inner unions i think it depends on what the union contains, if an intersection is involved I don't think you can simply flaten it? 

![image](https://user-images.githubusercontent.com/99813521/157265173-81b7d481-1d11-4307-9698-8da688a0f6ed.png)

Reply:
Why does the algorithm not return always things in the same order? is there a non-deterministic step somewhere?

As for the flattening, BGPs are not *exactly* intersections, are joins or even cartesian products. Anyhow, my thought is that:
Union(tp1, tp2, Union(BGP, tp3)) is the same as Union(tp1, tp2, BGP, tp3), right?

Reply:
The only step I can think of is the find function that is used to get the mappings. That might return the mappings in different orders, which in turn means that they are added in different orders to the union.

I guess to yes! Should I try to flatten out bthe results then and compare them?

Reply:
No, don't sort the mappings in the production code in order to make the unit test work. If the order of the subpatterns in the returned union pattern is nondeterministic, then that's fine; it is still a correct behavior. Therefore, the code of the unit test needs to be extended to accept this correct behavior (any order of the expected subpatterns is okay and should pass the test) instead of adding an expensive but unnecessary ordering process into the runtime code.